Resource arm analysis services server

r0bnet opened the pull request:
Created this resource as it was requested in an open issue:
https://github.com/terraform-providers/terraform-provider-azurerm/issues/1734
I omitted the `gateway_resource_id` and `backup_blob_container_uri` for now as i had problems in testing it (for gateway resource id i'm not even sure for it is needed).

There is also currently no acc test for the `admin_users` because i'm having difficulties of creating them in the test. Manual tests were successful though.

tlaur commented:
Brilliant, well done on working on this, will take a closer look and take this out for a test spin later.

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	}
+}
+
+func flattenAnalysisServicesServerSku(sku *analysisservices.ResourceSku) string {
```
I don't think this function is necessary.

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	return validation.StringInSlice(connectionModes, true)
+}
+
+func expandAnalysisServicesServerSku(sku string) *analysisservices.ResourceSku {
```
I don't think this function is necessary.

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	return firewallSettings.EnablePowerBIService, fwRules
+}
+
+func flattenAnalysisServicesServerQuerypoolConnectionMode(serverProperties *analysisservices.ServerProperties) *string {
```
I don't think this function is necessary.

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	return &firewallSettings
+}
+
+func flattenAnalysisServicesServerAdminUsers(serverAdministrators *analysisservices.ServerAdministrators) *[]string {
```
I don't think this function is necessary.

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	return &analysisservices.ServerAdministrators{Members: &members}
+}
+
+func expandAnalysisServicesServerQuerypoolConnectionMode(d *schema.ResourceData) *analysisservices.ConnectionMode {
```
I don't think this function is necessary.

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	analysisServicesServer := analysisservices.Server{
+		Name:             &name,
+		Location:         &location,
+		Sku:              expandAnalysisServicesServerSku(sku),
```
Given this is just .a string i don't think we need an expand function here

r0bnet replied:
removed (in case you mean we can do it inline)

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	log.Printf("[INFO] preparing arguments for Azure ARM Analysis Services Server creation.")
+
+	name := d.Get("name").(string)
+	resGroup := d.Get("resource_group_name").(string)
```
_minor_ Ideally we should call this `resourceGroup` for consistency

r0bnet replied:
renamed the variable in all places

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+
+	d.Set("sku", flattenAnalysisServicesServerSku(server.Sku))
```
This should simply be:
```suggestion
	d.Set("sku", server.Sku)
```

r0bnet replied:
done

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	tags := d.Get("tags").(map[string]interface{})
+
+	analysisServicesServer := analysisservices.ServerUpdateParameters{
+		Sku:                     expandAnalysisServicesServerSku(sku),
```
As above

r0bnet replied:
done

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+}
+
+func expandAnalysisServicesServerProperties(d *schema.ResourceData) *analysisservices.ServerProperties {
+	adminUsers := expandAnalysisServicesServerAdminUsers(d)
```
I don't think there is a need for this function

r0bnet replied:
i'm not sure if i understand correctly. IMHO this function does more than i want to have simply inline. Can you describe in detail?

tombuildsstuff replied:
it doesn't make much difference either way - so I think this is fine as-is

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	//	serverProperties.BackupBlobContainerURI = backupBlobContainerUri.(*string)
+	//}
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
```
I don't think there is a need for this function

r0bnet replied:
i'm not sure if i understand correctly. IMHO this function does more than i want to have simply inline. Can you describe in detail?

tombuildsstuff replied:
it doesn't make much difference either way - so I think this is fine as-is

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+
+	serverProperties.IPV4FirewallSettings = expandAnalysisServicesServerFirewallSettings(d)
+
+	if connectionMode := expandAnalysisServicesServerQuerypoolConnectionMode(d); connectionMode != nil {
```
I don't think there is a need for this function

r0bnet replied:
i'm not sure if i understand correctly. IMHO this function does more than i want to have simply inline. Can you describe in detail?

katbyte commented on go.sum:
```diff
@@ -16,6 +16,7 @@ github.com/Azure/azure-sdk-for-go v21.3.0+incompatible/go.mod h1:9XXNKU+eRnpl9mo
 github.com/Azure/azure-sdk-for-go v29.0.0+incompatible/go.mod h1:9XXNKU+eRnpl9moKnB4QOLf1HestfXbmab5FXxiDBjc=
 github.com/Azure/azure-sdk-for-go v30.0.0+incompatible h1:6o1Yzl7wTBYg+xw0pY4qnalaPmEQolubEEdepo1/kmI=
 github.com/Azure/azure-sdk-for-go v30.0.0+incompatible/go.mod h1:9XXNKU+eRnpl9moKnB4QOLf1HestfXbmab5FXxiDBjc=
+github.com/Azure/azure-sdk-for-go v30.0.0+incompatible/go.mod h1:9XXNKU+eRnpl9moKnB4QOLf1HestfXbmab5FXxiDBjc=
```
This looks like a spurious duplicate line?

r0bnet replied:
True. Merging error. Removed

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	if serverProps := server.ServerProperties; serverProps != nil {
+		d.Set("admin_users", *flattenAnalysisServicesServerAdminUsers(server.AsAdministrators))
+
+		//if serverProps.BackupBlobContainerURI != nil {
```
If this isn't required we should remove the comment

r0bnet replied:
removed

katbyte commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,449 @@
+	d.Set("sku", flattenAnalysisServicesServerSku(server.Sku))
+
+	if serverProps := server.ServerProperties; serverProps != nil {
+		d.Set("admin_users", *flattenAnalysisServicesServerAdminUsers(server.AsAdministrators))
```
Idealy we should be using serverprops instead of server

katbyte replied:
Also as this is a top level property we shouldn't need a flatten function

r0bnet replied:
> Idealy we should be using serverprops instead of server

don't understand completely. Can you go into more detail?

katbyte commented on azurerm/resource_arm_analysis_services_server_test.go:
```diff
@@ -0,0 +1,379 @@
+	resourceName := "azurerm_analysis_services_server.test"
+	ri := tf.AccRandTimeInt()
+
+	config1 := testAccAzureRMAnalysisServicesServer_firewallSettings(ri, testLocation(), true, make([]map[string]string, 0))
```
Instead of passing in parameters like this could we just have straight HCL configs instead?

katbyte commented on azurerm/resource_arm_analysis_services_server_test.go:
```diff
@@ -0,0 +1,379 @@
+	})
+}
+
+// Currently unsure how to test it as the email addresses need to exist in the AD
```
Would it be possible to create the user's via the azuread provider & use them here?

r0bnet replied:
Talked about this with @tombuildsstuff and he proposed to work with a env var which points to a email address in the azure ad. That okay for you? Then everyone can test in their own azure tenant.

r0bnet replied:
You now need to specify two email addresses as env variables to run this test. Unfortunately it's not working as expected -> test fails but i don't know why.

katbyte commented on azurerm/resource_arm_analysis_services_server_test.go:
```diff
@@ -0,0 +1,379 @@
+}
+
+resource "azurerm_analysis_services_server" "test" {
+  name                = "acctestass%d"
```
Could we ensure capitals work for the name here?
```suggestion
  name                = "acctestASS%d"
```

r0bnet replied:
renamed accordingly

r0bnet replied:
had to revert it as only lowercase is allowed

katbyte commented on azurerm/resource_arm_analysis_services_server_test.go:
```diff
@@ -0,0 +1,379 @@
+
+  tags = {
+	label = "test1"
+	env   = "prod"
```
Could we ensure capitals work here?
```suggestion
	ENV   = "Prod"
```

r0bnet replied:
changed to uppercase

r0bnet commented:
Thanks @katbyte. I currently don't have the possibility to check your annotations but will try to come back to it next week.

stefan-rapp commented:
Do you know when the creation of an Azure Analysis Service resource will be part of the next Version? Need this feature for a customer project? Thx

r0bnet commented:
So now i fixed the failing test and changed most of the stuff that i was told to.
@stefan-rapp we'll have to wait what the maintainers will say. Sorry for being so late. Hope this will be included in 1.33.0 at least.

stefan-rapp commented:
Hi @r0bnet, Nice! Thank you very much. Appreciate it! 👍

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+			//	Optional:     true,
+			//	Sensitive:    true,
+			//	ValidateFunc: validate.URLIsHTTPOrHTTPS,
+			//},
```
could we remove this for the moment?

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+
+	server, err := client.GetDetails(ctx, resourceGroup, name)
+	if err != nil {
+		return err
```
to make this more descriptive for users - could we wrap this error?

```suggestion
		return fmt.Errorf("Error retrieving Analytics Services Server %q (Resource Group %q): %+v", name, resourceGroup, err)
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+
+	d.Set("sku", *server.Sku.Name)
```
there's a potential crash here if this is nil/a bad API response; could we make this:

```
if server.Sku != nil {
  d.Set("sku", server.Sku.Name)
}
```

`d.Set` will handle the property being a pointer and de-reference it to an empty string if required

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+
+		enablePowerBi, fwRules := flattenAnalysisServicesServerFirewallSettings(serverProps)
+		d.Set("enable_power_bi_service", enablePowerBi)
+		d.Set("ipv4_firewall_rule", fwRules)
```
since this is a complex object could we check for errors when setting here?

```suggestion
		if err := d.Set("ipv4_firewall_rule", fwRules); err != nil {
			return fmt.Errorf("Error setting `ipv4_firewall_rule`: %s", err)
		}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+	}
+
+	fwRules = make([]interface{}, 0)
+	for _, fwRule := range *firewallSettings.FirewallRules {
```
if `firewallSettings.FirewallRules` is nil (for example, a bad server response) there's a potential crash here; as such could we make this:

```suggestion
	if firewallSettings.FirewallRules != nil {
		for _, fwRule := range *firewallSettings.FirewallRules {
		...
	}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+	fwRules = make([]interface{}, 0)
+	for _, fwRule := range *firewallSettings.FirewallRules {
+		output := make(map[string]interface{})
+		output["name"] = *fwRule.FirewallRuleName
```
there's a potential panic here if this is nil - as such could we nil-check around this?

```suggestion
		if fwRule.FirewallRuleName != nil {
			output["name"] = *fwRule.FirewallRuleName
		}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+	for _, fwRule := range *firewallSettings.FirewallRules {
+		output := make(map[string]interface{})
+		output["name"] = *fwRule.FirewallRuleName
+		output["range_start"] = *fwRule.RangeStart
```
(same here)

```suggestion
		if fwRule.RangeStart != nil {
			output["range_start"] = *fwRule.RangeStart
		}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+		output := make(map[string]interface{})
+		output["name"] = *fwRule.FirewallRuleName
+		output["range_start"] = *fwRule.RangeStart
+		output["range_end"] = *fwRule.RangeEnd
```
(same here)

```suggestion
		if fwRule.RangeEnd != nil {
			output["range_end"] = *fwRule.RangeEnd
		}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+
+	server, err := client.GetDetails(ctx, resourceGroup, name)
+	if err != nil {
+		return err
```
could we wrap this error to provide a better explanation to users if this fails?

```suggestion
		return fmt.Errorf("Error retrieving Analysis Services Server %q (Resource Group %q): %s", name, resourceGroup, err)
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+		return fmt.Errorf("Cannot read ID of Analysis Services Server %q (Resource Group %q)", name, resourceGroup)
+	}
+
+	d.SetId(*server.ID)
```
since the ID shouldn't have changed during the Update (and this is a separate method) could we remove this (and the nil-check above?)

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,406 @@
+		firewallSettings.EnablePowerBIService = utils.Bool(enablePowerBi.(bool))
+	} else {
+		firewallSettings.EnablePowerBIService = utils.Bool(false)
+	}
```
since the default value for this will be `false` we can just default this above:

```suggestion
	firewallSettings := analysisservices.IPv4FirewallSettings{
		EnablePowerBIService: utils.Bool(d.Get("enable_power_bi_service").(bool))
	}
```

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server_test.go:
```diff
@@ -0,0 +1,409 @@
+// ARM_ACC_EMAIL1 and ARM_ACC_EMAIL2 must be set and existing emails in the tenant's AD to work properly
+func TestAccAzureRMAnalysisServicesServer_adminUsers(t *testing.T) {
+	const ArmAccEmail1 = "ARM_ACC_EMAIL1"
+	const ArmAccEmail2 = "ARM_ACC_EMAIL2"
```
could we make these keys more descriptive? perhaps `ARM_ACCTEST_ADMIN_EMAIL1` and `ARM_ACCTEST_ADMIN_EMAIL2`?

r0bnet commented:
@tombuildsstuff resolved all the issues :)

tombuildsstuff commented:
@r0bnet restarted travis :)

tombuildsstuff commented on azurerm/resource_arm_analysis_services_server.go:
```diff
@@ -0,0 +1,403 @@
+}
+
+func flattenAnalysisServicesServerFirewallSettings(serverProperties *analysisservices.ServerProperties) (enablePowerBi *bool, fwRules []interface{}) {
+	if serverProperties.IPV4FirewallSettings == nil {
```
there's a potential crash here if `serverProperties` is nil - so could we make this:

```suggestion
	if serverProperties == nil || serverProperties.IPV4FirewallSettings == nil {
```

tombuildsstuff commented on website/docs/r/analysis_services_server.html.markdown:
```diff
@@ -0,0 +1,84 @@
+---
```
could we add this new resource [to the sidebar](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/website/azurerm.erb#L389-L398)?

tombuildsstuff commented:
@r0bnet I hope you don't mind but the tests identified an issue so I pushed a commit to fix that; but the tests now pass / this LGTM 👍 

<img width="534" alt="Screenshot 2019-07-22 at 10 38 49" src="https://user-images.githubusercontent.com/666005/61618325-e5998600-ac6c-11e9-95e6-6de071113e8a.png">
